### PR TITLE
feat(download): 🔊 Improve and standardize logging for downloaded configs.

### PR DESCRIPTION
**Pull request by Laubi**
#### What this PR does / Why we need it:
For configs and settings we're no longer logging when there has been nothing downloaded. For buckets and automation resources, it's still logged to make it very prominent.

Additionally, this PR standardizes the logging a bit more.

Log examples:
<details><summary>Some logs with debug enabled:</summary>

```
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:problem.notifications
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:process.built-in-process-monitoring-rule"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:dashboards.presets
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:rum.web.custom-errors"
2023-12-14T14:02:48+01:00       info    Downloaded 114 settings for schema "builtin:tags.auto-tagging".
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:rum.overload-prevention
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:monitored-technologies.go"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:process-group.advanced-detection-rule
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:synthetic.synthetic-availability-settings"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:dashboards.general
2023-12-14T14:02:48+01:00       info    Downloaded 2 settings for schema "builtin:synthetic.http.performance-thresholds".
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:anomaly-detection.kubernetes.namespace
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:rum.mobile.request-errors"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:failure-detection.environment.parameters
2023-12-14T14:02:48+01:00       info    Downloaded 1 settings for schema "builtin:unified-services-enablement".
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:ownership.teams"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:rum.ip-mappings
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:anomaly-detection.infrastructure-aws
2023-12-14T14:02:48+01:00       info    Downloaded 2 settings for schema "builtin:synthetic.browser.outage-handling".
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:audit-log
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:bizevents.http.incoming"
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:disk.analytics.extension"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:rum.web.key-performance-metric-custom-actions
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:rum.web.ipaddress-exclusion
2023-12-14T14:02:48+01:00       debug   Did not find any settings to download for schema "builtin:logmonitoring.log-custom-attributes"
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:bizevents-processing-buckets.rule
2023-12-14T14:02:48+01:00       debug   Downloading all settings for schema builtin:cloud.kubernetes
```
</details>

<details><summary>Without debug enabled</summary>

```
2023-12-14T14:02:31+01:00       info    Downloaded 1 settings for schema "builtin:sessionreplay.web.privacy-preferences".
2023-12-14T14:02:31+01:00       info    Downloaded 1 settings for schema "builtin:service-detection.full-web-service".
2023-12-14T14:02:31+01:00       info    Downloaded 53 settings for schema "builtin:span-attribute".
2023-12-14T14:02:31+01:00       info    Downloaded 2 settings for schema "builtin:synthetic.browser.scheduling".
2023-12-14T14:02:31+01:00       info    Downloaded 2 settings for schema "builtin:alerting.profile". Skipped persisting 2 unmodifiable setting(s).
```
</details> 

<details><summary>Workflows/buckets w/o debug nabled</summary>

```
2023-12-14T14:07:52+01:00       info    Downloaded 3 objects for workflow
2023-12-14T14:07:52+01:00       error   Failed to fetch all objects for automation resource business-calendar: API request HTTP GET https://xxxxxxxx.apps.dynatracelabs.com/platform/automation/v1/business-calendars?adminAccess=true&offset=0 failed with status code 403: {"error":{"message":"Only admins can request admin access.","code":403}}
2023-12-14T14:07:52+01:00       error   Failed to fetch all objects for automation resource scheduling-rule: API request HTTP GET https://xxxxxxx.apps.dynatracelabs.com/platform/automation/v1/scheduling-rules?adminAccess=true&offset=0 failed with status code 403: {"error":{"message":"Only admins can request admin access.","code":403}}
2023-12-14T14:07:52+01:00       info    Downloading Grail buckets
2023-12-14T14:07:52+01:00       info    Downloaded 5 buckets. Skipped persisting 12 unmodifiable bucket(s).
```
</details>

### CODE REVIEW

File: internal/log/field/field.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ func Coordinate(coordinate coordinate.Coordinate) Field {
 
 // Type builds a Field containing information about a config type. This is used in cases where no full coordinate exists,
 // but only a config type is known - for example in download or deletion
-func Type(t string) Field {
+func Type[X ~string](t X) Field {
```
Comment (UnseenWizzard) [inline internal/log/field/field.go:58]: next-level-nit: I'd have preferred a dedicated commit for this change that also tells me WHY this change happened. At a glance/without looking at things like automation.Resource being a dedicated 'string' type elsewhere this looks like a pretty strange change

File: pkg/download/automation/automation.go [hunk 2]
```diff
@@ -64,39 +64,44 @@ func (d *Downloader) Download(projectName string, automationTypes ...config.Auto
 
 	configsPerType := make(v2.ConfigsPerType)
 	for _, at := range automationTypes {
+		lg := log.WithFields(field.Type(at.Resource))
```
Comment (UnseenWizzard) [inline pkg/download/automation/automation.go:67]: nit/note: This actually increased the memory footprint of the logger, as you now wrap the log core twice, once for the resource filed, once again for the extra "WithFields()" every log message uses. 
WithFields(<type>, <error>, <...>) will only be one wrap

File: pkg/download/automation/automation.go [hunk 3]
```diff
@@ -64,39 +64,44 @@ func (d *Downloader) Download(projectName string, automationTypes ...config.Auto
...
+		lg.WithFields(field.F("configsDownloaded", len(objects))).Info("Downloaded %d objects for %s", len(objects), string(at.Resource))
 
 		var configs []config.Config
 		for _, obj := range objects {
 
 			configId := obj.ID
 
 			if escaped, err := escapeJinjaTemplates(obj.Data); err != nil {
-				log.WithFields(field.Coordinate(coordinate.Coordinate{Project: projectName, Type: string(at.Resource), ConfigId: configId}), field.Error(err)).Warn("Failed to escape automation templating expressions for config %v (%s) - template needs manual adaptation: %v", configId, at.Resource, err)
+				lg.WithFields(field.Coordinate(coordinate.Coordinate{Project: projectName, Type: string(at.Resource), ConfigId: configId}), field.Error(err)).Warn("Failed to escape automation templating expressions for config %v (%s) - template needs manual adaptation: %v", configId, at.Resource, err)
```
Comment (UnseenWizzard) [inline pkg/download/automation/automation.go:104]: this log has a `type` and `coord` field now

File: pkg/download/automation/automation.go (hunk 2 above)
Comment (Laubi) [reply pkg/download/automation/automation.go:67]: Is this noticeable that we should work on that? As I see it, the loggers here are so short-lived that it should not be an issue, right?

File: pkg/download/classic/downloader.go [hunk 4]
```diff
@@ -109,32 +109,37 @@ func (d *Downloader) downloadAPIs(apisToDownload api.APIs, projectName string) p
...
 			}
 
-			if remoteCount == len(cfgs) {
-				log.WithFields(field.Type(currentApi.ID)).Info("Downloaded %d configurations for API %q", len(cfgs), currentApi.ID)
-			} else {
-				log.WithFields(field.Type(currentApi.ID)).Info("Downloaded %d configurations for API %q. Skipped persisting %d unmodifiable config(s).", len(cfgs), currentApi.ID, remoteCount-len(cfgs))
+			lg = lg.WithFields(field.F("configsDownloaded", len(cfgs)))
+			switch remoteCount {
+			case 0:
+				lg.WithFields().Debug("Did not find any configurations to download for API %q", currentApi.ID)
```
Comment (UnseenWizzard) [inline pkg/download/classic/downloader.go:138]: empty WithFields

File: pkg/download/bucket/bucket.go [hunk 5]
```diff
@@ -75,24 +75,34 @@ func (d *Downloader) Download(projectName string, _ ...config.BucketType) (v2.Co
...
 			continue
 		}
 		result = append(result, c)
 	}
 
-	log.Info("Downloaded %d Grail buckets", len(result))
+	lg = lg.WithFields(field.F("configsDownloaded", len(result)))
+	switch len(objects) {
+	case 0:
+		// Info on purpose. Most types have a lot of objects, so skipping printing 'not found' in the default case makes sense. Here it's kept on purpose as bucket is only one type.
```
Comment (UnseenWizzard) [inline pkg/download/bucket/bucket.go:99]: This makes me wonder when we should log this as debug and when as info. 
From a user perspective I think it's less dependent on the amount of types there are, but rather if I had requested the type. 

E.g. if I wanted to download a specific setting, I'd probably find an "info No settings found" helpful. 
If I downloaded everything, I don't think I'd need different levels of information for buckets or other types.

Just some thoughts to discuss, and most certainly not a change to do as part of this PR!

Comment (Laubi) [reply pkg/download/bucket/bucket.go:99]: I had the same thoughts as you're having. If I want to download the schema 'x', but no configs are available, I might want to see a message. For now opted to go this way, as if I'm not sure if it works, users can still enable the debug mode. 

Imo, when we rewrite the download code at some point, we really should include that use-case. However, we would need to restructure our download code immensely to handle that, so currently it's out of scope.

File: pkg/download/automation/automation.go (hunk 2 above)
Comment (UnseenWizzard) [reply pkg/download/automation/automation.go:67]: More a question of why you moved it into a shared partially initialized logger struct. If you did it due to the overhead of extra logger cores (which we had some issues with in the past), it would not actually have that effect.
If just for personal preference of structure, it's perfectly fine here IMO. 
Thus a nit/note, just to let you know :)

File: pkg/download/automation/automation.go (hunk 3 above)
Comment (UnseenWizzard) [reply pkg/download/automation/automation.go:104]: I really just see it as redundant, especially with the doc comment of field.Type saying it's used when no full coordinate is available

**Review by UnseenWizzard — APPROVED** [review]
